cloudferrylib/os/estimation/procedures.py

```diff
@@ -33,25 +35,26 @@ def estimate_copy(cloud_name, tenant):
         accounted_volumes = set()
         accounted_images = set()
 
-        for server in list_filtered(session, nova.Server, cloud_name, tenant):
-            for ephemeral_disk in server.ephemeral_disks:
-                total_ephemeral_size += ephemeral_disk.size
-            if server.image is not None \
-                    and server.image.object_id not in accounted_images:
-                total_image_size += server.image.size
-                accounted_images.add(server.image.object_id)
-            for volume in server.attached_volumes:
-                if volume.object_id not in accounted_volumes:
-                    total_volume_size += volume.size
-                    accounted_volumes.add(volume.object_id)
-        for volume in list_filtered(session, cinder.Volume, cloud_name,
-                                    tenant):
-            if volume.object_id not in accounted_volumes:
-                total_volume_size += volume.size
-        for image in list_filtered(session, glance.Image, cloud_name, tenant):
-            if image.object_id not in accounted_images:
-                total_image_size += image.size
+        for obj in migration.query.search(session, migration.source):
+            if isinstance(obj, nova.Server):
+                for ephemeral_disk in obj.ephemeral_disks:
+                    total_ephemeral_size += ephemeral_disk.size
+                if obj.image is not None \
+                        and obj.image.object_id not in accounted_images:
+                    total_image_size += obj.image.size
+                    accounted_images.add(obj.image.object_id)
+                for volume in obj.attached_volumes:
+                    if volume.object_id not in accounted_volumes:
+                        total_volume_size += volume.size
+                        accounted_volumes.add(volume.object_id)
+            elif isinstance(obj, cinder.Volume):
+                if obj.object_id not in accounted_volumes:
+                    total_volume_size += obj.size
+            elif isinstance(obj, glance.Image):
+                if obj.object_id not in accounted_images:
+                    total_image_size += obj.size
```

Comment:
I fear that just implementing this functionality as methods of cinder.Volume, glance.Image and nova.Server will only complicate things; we will leak implementation details of estimating migration into this classes.
I will try to implement visitor pattern, but I'm not sure that it will be cleaner.